packages/api-derive/src/type/util.ts

```diff
@@ -4,23 +4,29 @@
 import type { AccountId, Digest } from '@polkadot/types/interfaces';
 
 export function extractAuthor (digest: Digest, sessionValidators: AccountId[] = []): AccountId | undefined {
-  const [pitem] = digest.logs.filter(({ type }) => type === 'PreRuntime');
+  let author;
```

Comment:
Not sure I would swap the order around - the most-used is `PreRuntime` (and most recent), so it should go first.

Additionally we have lost the `return engine.extractAuthor(data, sessionValidators);` from the citems now. In reality the specific log for Moonbeam should probably be in there?

Reply:
>Not sure I would swap the order around - the most-used is PreRuntime (and most recent), so it should go first

If you look at the logs in https://github.com/PureStake/moonbeam/pull/159#discussion_r560418535 you'll see that Moonbeam also has a PreRuntime log but it doesn't work. So we should first check for Consensus before we look at PreRuntime.

>Additionally we have lost the return engine.extractAuthor(data, sessionValidators); from the citems now. In reality the specific log for Moonbeam should probably be in there?

`return engine.extractAuthor(data, sessionValidators)` doesn't work for Moonbeam. Since all other uses of Consensus have been removed, I assumed it would be simple to just drop it. I can reestablish it for non-h160 addresses if you think it's useful

Reply:
> Moonbeam also has a PreRuntime log

That's only true when you're talking about the standalone node with Aura. The parachain node (the primary moonbeam product) does not have a PreRuntime digest, although it may in the future). For example, look at the live testnet.

![image](https://user-images.githubusercontent.com/2915325/105712988-96c74f00-5ee8-11eb-88ed-11f5cb2265e6.png)

Reply:
For instance, by just returning here we have actually broken Kulupu which also uses the Consensus digest.

The actual extraction needs to happen here - https://github.com/polkadot-js/api/blob/master/packages/types/src/generic/ConsensusEngineId.ts

(Aura, Babe & Kulupu POW is done there)

Reply:
TL;DR this logic needs to be kept in-tact, it will determine the type of digest and defer to the `ConsensusEngineId` for extracting.

Reply:
But if we keep this logic, this won't work for the standalone, which would have a PreRuntime log, but where the relevant information is in the Consensus log

Reply:
The logic is as follow, directl from the source, commented here ...

```js
  // try to find any PreRuntime digests here  
  const [pitem] = digest.logs.filter(({ type }) => type === 'PreRuntime');

  if (pitem) {
   // if we have some PreRuntime logs, follow this path
    ...
  } else {
    // we didn't have any of the previous type, now try to find Consensus logs
    const [citem] = digest.logs.filter(({ type }) => type === 'Consensus');
    ...
  }
```

So if doesn't require PreRuntime, it is one of the two.

Reply:
I'm happy to swap the oder above, as long as we don't try and make the extraction here as well.